feat(https): use https by default

ericdeansanchez opened the pull request:
## feat(https): use https by default

### Description
This PR changes the default behavior of the single-argument
`URLBuilder` constructor. This change ensures that a default
constructed `URLBuilder` has its `useHttps` member variable
set to `true` so that urls are generated for the https scheme.

### Change (Breaking)
__This change in behavior represents a breaking change__. 

### Before 
Prior to this change, users could rely on a default-constructed `URLBuilder`'s
`useHttps` member variable to be `false`. This meant that generated urls
would have their scheme set to "http" by default.

### Now
**Now, `URLBuilder`'s `useHttps` member variable is set to `true`**. This means
that generated urls now have their scheme set to "https".

### Testing
-  three new tests have been added
-  one to test the default constructor (single-parameter)
    constructs an https-url by default (by calling `createUrl`)
  - [x] this new test __does fail__ before the breaking change
  - [x] this new test __does pass__ after this breaking change
- one to test that `setUseHttps` correctly sets `false`
- one to test that `setUseHttps` correctly sets `true`

### Misc
In the [initial commit](https://github.com/imgix/imgix-java/commit/93dacc10ee90083bab1aee8e6e293cd23d40ec1c), IntelliJ tried to sneak in some spelling suggestions.

The suggested change was erroneously pushed with the above commit. IntelliJ was
right in that a change could/should be made.

An extra `a` _does_ improve the grammar of the exception message of which it is now part of.

We also took this opportunity to reformat the code base to indent files with spaces instead of tab characters (for more consistency).

ericdeansanchez committed:
feat(https): use https by default

This PR changes the default behavior of the single-argument
`URLBuilder` constructor. This change ensures that a default
constructed `URLBuilder` has its `useHttps` member variable
set to `true` so that urls are generated for the https scheme.

__This change in behavior represents a breaking change__. Prior to this
change, users could rely on a default-constructed `URLBuilder`'s
`useHttps` member variable to be `false`.

The intention was to test this change. However, with `useHttps` set
to `private` and without reflection, this cannot currently be done.

Here are a few options:
- a new getter could be defined, i.e. `getUseHttps` which feels a little
  clumsy at the moment, __or__
- the interface must be refactored, __or__
- this feature doesn't get tested (unless there's some other way)
  - not-testing is a decision that should be made after deliberation
  - not-testing should only be an option if there are no other
    _reasonable_ options
  - tests before this __breaking change__ should have _passed_
  - tests after this __breaking change__ should have _failed_
  - then a new test should have been written

In order to keep this change appropriately scoped (and modular), any
testing of this feature will be done:

- in a separate PR entirely (possibly after adding a getter or
  refactoring, i.e. private -> public member), _and_
- after discussing potential changes with the SDK Team

ericdeansanchez committed:
style(grammar): correct grammar for exception msg

sherwinski commented on src/main/java/com/imgix/URLBuilder.java:
```diff
@@ -31,8 +31,15 @@ public URLBuilder(String domain, boolean useHttps, String signKey, boolean inclu
+	 * to `true` by default.
+	 *
+	 * @param domain - a valid domain string, e.g. `example.imgix.net`
+	 */
```
Also, if we're going to add doc comments for one of the constructors, it only feels right to add it for all other public methods as well.

ericdeansanchez replied:
This is true, and desired. I will document everything else in a separate PR. I wanted to keep this change scoped appropriately.

sherwinski replied:
Great, in that case let's remove it here for cleanliness and include it in your separate PR.

sherwinski reviewed (changes requested):
Looks good, but I think it would be sufficient to test the output of `createURL` behaves as expected with the two different `useHttps` scenarios: the default `true` and when overriden with `false`.
Those tests can be included in this PR as they pertain to this change.

ericdeansanchez committed:
feat(https): use https by default

This PR changes the default behavior of the single-argument
`URLBuilder` constructor. This change ensures that a default
constructed `URLBuilder` has its `useHttps` member variable
set to `true` so that urls are generated for the https scheme.

__This change in behavior represents a breaking change__. Prior to this
change, users could rely on a default-constructed `URLBuilder`'s
`useHttps` member variable to be `false`.

~~The intention was to test this change. However, with `useHttps` set
to `private` and without reflection, this cannot currently be done.~~
**Edit:** it can be done by testing the functionality through the
`createUrl` method.

Here are a few options:
- a new getter could be defined, i.e. `getUseHttps` which feels a little
  clumsy at the moment, __or__
- the interface must be refactored, __or__
- this feature ~~doesn't get tested~~ (unless there's some other way)
  - not-testing is a decision that should be made after deliberation
  - not-testing should only be an option if there are no other
    _reasonable_ options
  - tests before this __breaking change__ should have _passed_
  - ~~tests after this __breaking change__ should have _failed_~~
  - ~~then a new test should have been written~~
  - a new test __has__ been added, however, it tests that `createUrl`
    produces a https-url, given a default-constructed url
  - [x] this new test _does fail_ before the breaking change
  - [x] this new test does _pass_ after this breaking change

~~In order to keep this change appropriately scoped (and modular), any
testing of this feature will be done:~~

- ~~in a separate PR entirely (possibly after adding a getter or
  refactoring, i.e. private -> public member), _and_~~
- [x] after discussing potential changes with the SDK Team

ericdeansanchez committed:
test:  uses https by default

sherwinski commented on src/test/java/com/imgix/test/TestAll.java:
```diff
@@ -28,6 +28,16 @@ public void testURLBuilderRaisesExceptionOnNoDomains() {
+		assertNotEquals("https://example.imgix.net/?ixlib=java-2.1.1", ub.createURL(""));
+		assertEquals("http://example.imgix.net/?ixlib=java-2.1.1", ub.createURL(""));
+	}
+
```
1) Let's not pass just an empty string to `createURL`, the tests should ideally mimic actual user behavior.
2) It looks like this one test is testing two separate things. But it also made me realize that we don't have a test for `setUseHttps`, which we should. So let's break it apart into 3 cases:

-  `testURLBuilderUsesHttpsByDefault()`: just L33-34
-  `testURLBuilderOverridesHttpsByDefault()`: using `new URLBuilder("example.imgix.net", false);`
-  `testsetUseHttps()`: L36

ericdeansanchez replied:
1. Okay, I saw passing an empty string as the "base case." Passing an empty string is allowable, and _could_ mimic user behavior, no?
2. I agree. This test tests way more than `useHttps`. This  test is testing more than two separate things. `createUrl(_)` calls `createUrl(_, _, _)`,  which in turn calls `URLHelper`. It is essentially an integration test at this point.
3. This is kind of what I meant by, "this behavior isn't easy to test [in isolation] given the current API" (paraphrasing).
4. I like the idea of splitting this up, what do you think about this?

```java
@Test
public void testURLBuilderUsesHttpsByDefault() {
}

@Test
public void testSetUseHttpsFalse() {
}

@Test
public void testSetUseHttpsTrue() {
}
```

sherwinski replied:
1. Yes it's allowable, but it would result in a URL with no path and therefore wouldn't actually resolve to an image.
2. These tests look good, please include an additional one for passing in `false` to the constructor.

ericdeansanchez replied:
1. Okay, I will add a path. However, we are already testing "too much" and I thought that passing an empty string was closest to testing the most basic url that can be default-constructed (since the interface does not allow for that kind of thing explicitly).  Would you like me to test a url and path that resolve to an actual image?
2. Got it, but why stop at the two-parameter constructor? There's still the 3 and 4-parameter constructors to handle.
3. Since every constructor eventually calls the 4-parameter constructor, I can write a test for that and we can go from there (or write tests for each constructor, but that leads me to point 4 below).
4. That being said, before we grow our test suite we should discuss a few more things offline (i.e. idiomatic use of `assertEquals` is `assertEquals(expected, actual)` right now many appear to be flipped).

ericdeansanchez committed:
feat(https): use https by default

sherwinski reviewed (approved):
LGTM 👍 

I just noticed it with all the reformatting but we should eventually remove `examples/com/imgix/examples/CycleDomainSharding.java` and `examples/com/imgix/examples/DomainSharding.java` from the examples. They refer to features that were deprecated/removed in an earlier release.
